src/crimson/net/ProtocolV2.cc

```diff
@@ -0,0 +1,1603 @@
+          logger().debug("{} received reconnect ok:"
+                         "sms={}, lossy={}, features={}",
+                         conn, reconnect_ok.msg_seq(),
+                         connect_seq, conn.policy.lossy, conn.features);
+          // TODO
+          // discard_requeued_up_to()
+          // backoff = utime_t();
+          return dispatcher.ms_handle_connect(
+              seastar::static_pointer_cast<SocketConnection>(
+                conn.shared_from_this()))
+          .handle_exception([this] (std::exception_ptr eptr) {
+            logger().error("{} ms_handle_connect caust exception: {}", conn, eptr);
+          });
+        }).then([this] {
+          return true;
+        });
+      default: {
+        unexpected_tag(tag, conn, "post_client_reconnect");
+        // won't be executed
+        return seastar::make_ready_future<bool>(false);
+      }
+    }
+  });
+}
+
+void ProtocolV2::execute_connecting()
+{
+  trigger_state(state_t::CONNECTING, write_state_t::delay, true);
+  seastar::with_gate(pending_dispatch, [this] {
+      enable_recording();
+      global_seq = messenger.get_global_seq();
+      return Socket::connect(conn.peer_addr)
+        .then([this](SocketFRef sock) {
+          socket = std::move(sock);
+          if (state == state_t::CLOSING) {
+            return socket->close().then([this] {
+              logger().info("{} is closed during Socket::connect()", conn);
+              abort_in_fault();
+            });
+          }
+          return seastar::now();
+        }).then([this] {
+          return banner_exchange();
+        }).then([this] (entity_type_t _peer_type,
+                        entity_addr_t _peer_addr) {
+          if (conn.peer_type != _peer_type) {
+            logger().debug("{} connection peer type does not match what peer advertises {} != {}",
+                           conn, conn.peer_type, (int)_peer_type);
+            dispatch_reset();
+            abort_in_close();
+          }
+        }).then([this] {
+          return client_auth();
+        }).then([this] {
+          if (!server_cookie) {
+            ceph_assert(connect_seq == 0);
+            return client_connect();
+          } else {
+            ceph_assert(connect_seq > 0);
+            // TODO: lossless policy
+            ceph_assert(false);
+            return client_reconnect();
+          }
+        }).then([this] (bool proceed_or_wait) {
+          if (proceed_or_wait) {
+            execute_ready();
+          } else {
+            execute_wait();
+          }
+        }).handle_exception([this] (std::exception_ptr eptr) {
+          // TODO: handle fault in CONNECTING state
+          return fault();
+        });
+    });
+}
+
+// ACCEPTING state
+
+seastar::future<> ProtocolV2::_auth_bad_method(int r)
+{
+  // _auth_bad_method() logic
+  ceph_assert(r < 0);
+  std::vector<uint32_t> allowed_methods;
+  std::vector<uint32_t> allowed_modes;
+  messenger.auth_server->get_supported_auth_methods(
```

Comment:
i'd suggest return a tuple or a pair of `std::vector<uint32_t>` instead of passing pointers as output parameter.

Reply:
fixed in https://github.com/ceph/ceph/pull/26710/commits/8c076fd351f76c0022570c3a6c8bfe726ad2c025